### PR TITLE
[WIP] Supporting  Contao 3 and Contao 4

### PR DESCRIPTION
**Pull request by aschempp**
This PR aims to fix https://github.com/contao/core-bundle/issues/278

The example extension can be found at https://github.com/aschempp/composer-test/blob/master/composer.json. The requirements are set on `contao/core-bundle` (or `contao/news-bundle` etc.) 

I had to make slight adjustments to the [composer-client](https://github.com/aschempp/composer-client/commits/develop) just to make it allow to install the develop-version of the plugin.

### In Contao 3
 1. Add Composer manager through the backend (ER2). In my case it was a Contao 3.2.5
 2. Go to command line to perform migration  
 2. Add `repositories` for my fork of `composer-plugin`, [composer-client](https://github.com/aschempp/composer-client/commits/develop) and [composer-test](https://github.com/aschempp/composer-test) in your `composer.json`
 3. Installing and uninstalling `aschempp/composer-test` works from both backend and command line

Performing migration in the backend caused a 30-sec timeout for me, because it tries to pull `contao/core-bundle` and all Symfony dependencies. That is however an existing problem of the Contao composer client, because it does install dependencies of `contao/core` (or `contao/core-bundle`) before the plugin is active and can stop it. We need to find a new workaround for that.

### In Contao 4
 1. Clean installation of Contao 4 using  
     `composer create-project contao/standard-edition Contao4 4.*@dev`
 2. Add `repositories` for my fork of `composer-plugin`, [composer-client](https://github.com/aschempp/composer-client/commits/develop) and [composer-test](https://github.com/aschempp/composer-test) in your `composer.json`
 3. Installing and uninstalling `aschempp/composer-test` works on the command line
 4. For the first time (the first extension requiring the plugin), you have to `composer update` several times. This is a known issue to the existing composer plugin.

## ToDos
 - [ ] Rewrite the plugin and other classes to use the Environment instances directly
 - [ ] Fix existing unit tests that rely on removed methods
 - [ ] See how we can support unit tests (e.g. on Travis) for Contao 3
 - [ ] Make sure `contao/core` and all other "provides" are disallowed when requested by an extension

--
/cc @discordier @leofeyer @Toflar

### CODE REVIEW

File: src/ContaoCommunityAlliance/Composer/Plugin/Environment/ContaoEnvironmentFactory.php [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    private function isContao3($rootDir)
+    {
+        return is_file($rootDir . '/system/config/constants.php');
+    }
+
+    private function isContao4($rootDir)
+    {
+        return is_file($rootDir . '/app/console');
```
Comment (leofeyer) [inline src/ContaoCommunityAlliance/Composer/Plugin/Environment/ContaoEnvironmentFactory.php]: You should probably check for `$rootDir . '/vendor/contao/core-bundle` instead, because you don't know if the final app contains the `app/console` script.

Comment (aschempp) [inline src/ContaoCommunityAlliance/Composer/Plugin/Environment/ContaoEnvironmentFactory.php]: It's hacky, but we do need the console to find the upload path (using `debug:container`).

Comment (leofeyer) [inline src/ContaoCommunityAlliance/Composer/Plugin/Environment/ContaoEnvironmentFactory.php]: What if the file is not present in the application? If someone uses the Symfony 3 directory structure, it will be `bin/console`.

Comment (aschempp) [inline src/ContaoCommunityAlliance/Composer/Plugin/Environment/ContaoEnvironmentFactory.php]: The whole thing is nowhere failure-proof (yet) ;-)
It also does not correctly parse the version if you would have `dev-develop` installed. It's more like a proof of concept.

Comment (leofeyer) [inline src/ContaoCommunityAlliance/Composer/Plugin/Environment/ContaoEnvironmentFactory.php]: > It's hacky, but we do need the console to find the upload path

What about parsing the `parameters.yml` file?

Comment (aschempp) [inline src/ContaoCommunityAlliance/Composer/Plugin/Environment/ContaoEnvironmentFactory.php]: That's not gonna work. First of all, the `uploadPath` is not set in `parameters.yml`. It can be set in `app/config.yml`, but it can also be set anywhere and imported from anywhere. And the default value is set by our bundle. The only really working thing is checking the container.

We might be able to use our own entry point or our own way of booting the kernel, but I don't know how reliable that would be (e.g. someone can "theoretically" move the `AppKernel.php` to any other place).

**Comment by aschempp** [comment]
I did rewrite A LOT on the plugin today. I think I now have also completely understand how it works and what it's supposed to do to solve our problem (regarding installation). Explained in `composer.json` terms, it's fairly simply:

 1. It will add a virtual package to composer, telling it that `contao/core` is already installed in the version of the available Contao installation.
 2. This `contao/core` package now has a section that says

        "replace": {
            "contao/core-bundle": self.version,
            "contao/news-bundle": self.version,
            "contao/calendar-bundle": self.version
            /* … all contao/contao 4 bundles */
        }

 3. It virtually adds a dependency to the installation to require the known version of the installation for `contao/core` and all bundles.

Providing the packages will composer tell not to install anything if the given version is compatible with the requirements of a to-be-installed package. Having the requirement in the root will prevent composer from trying to update any of these packages, but fail to add the to-be-installed package.

For existing packages, adding `"require": { "contao/core": >3.2,<4 }` will work perfectly if Contao 3.4 is installed.

For packages supporting Contao 3 and 4, the requirement must be `"require": { "contao/core-bundle": >3.2,<5 }`. In Contao 3, the plugin will take care of "providing" `contao/core-bundle`. In Contao 4, it's actually available in composer (the plugin does not provide any package information on Contao 4).

File: src/ContaoCommunityAlliance/Composer/Plugin/Environment/Contao3Environment.php [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+            $constantsFile = $this->rootDir . '/system/config/constants.php';
+
+            if (!file_exists($constantsFile)) {
+                throw new ConstantsNotFoundException('Could not find constants.php in ' . $this->rootDir);
+            }
+
+            $contents = file_get_contents($constantsFile);
+
+            if (!preg_match('#define\(\'VERSION\', \'([^\']+)\'\);#', $contents, $match)) {
+                throw new ConstantsNotFoundException('Could not find the Contao build.');
```
Comment (leofeyer) [inline src/ContaoCommunityAlliance/Composer/Plugin/Environment/Contao3Environment.php]: The message should be "Could not find the Contao version", shouldn't it?

Comment (aschempp) [inline src/ContaoCommunityAlliance/Composer/Plugin/Environment/Contao3Environment.php]: Fixed in a64da74

**Comment by aschempp** [comment]
How do we want to continue here? If there is no feedback from anyone, we can't continue with the development. 

I'm not familiar with the CCA organisation, is the organisation supposed to support this repo? Otherwise if there is noone available to maintain it, maybe it should be transferred?

/cc @xtra @tristanlins

**Comment by tristanlins** [comment]
I'm out of scope here :-/

/fw @xtra

**Comment by discordier** [comment]
@aschempp Sorry I had no time to make indepth checks and tests on your approach.
As we discussed on mumble while you were initially writing this, I am still not convinced that all the corner cases we solved in the past will still be covered.

The overall picture I see here, as described in your comment before, is pretty good and nifty. Yet it must be polished.

Now a little offtopic
On the generic question about maintaining this, yes the CCA is maintaining it. But as you might have already noticed, we are a bunch of Contao enthusiasts dedicating our spare time to projects for the public. We maintain projects because "there is a need for them" or because they need an incubation home.
The bad part of the story is, that aside from @stefanheimes and myself, we have pretty few people in the CCA that are able to code but way too many (important) projects that would need more manpower. Therefore it simply takes time to check all PRs and issues, I bet you know that. :)

**Comment by aschempp** [comment]
So how should we continue here? My PR is supposed to show that my attempt works. Would it make sense to get the composer plugin developed by the Contao core team? Because it is pretty essential for all extension developers and therefore the success of Contao imho.

**Comment by discordier** [comment]
I have rebased your PR and fixed the unit test as good as I could.
See [contao4](https://github.com/contao-community-alliance/composer-plugin/tree/contao4) branch.

The contao4 environment however is not tested at all yet.
